marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.smart-mastering/survivorship/merging/options.xqy

```diff
@@ -564,3 +583,300 @@ declare function merge-impl:propertyspec-to-json($property-spec as element()) as
     json:transform-to-json($transformed-xml, $config)/*
 };
 
+declare variable $_cached-compiled-merge-options as map:map := map:map();
+
+declare function merge-impl:compile-merge-options(
+  $merge-options as item() (: as node()|json:object :)
+) {
+  merge-impl:compile-merge-options(
+    $merge-options,
+    fn:false()
+  )
+};
+
+(:
+ : Calculate merge information once per unique merge options in request to reduce repeat logic
+ : @param $merge-options  Options specifying how documents will be merged
+ : @param $only-warn-on-error  boolean indicating if errors should be returned rather than thrown
+ : @return map:map with compiled information about merge options
+ :)
+declare function merge-impl:compile-merge-options(
```

Comment:
I'm wondering if this function can be broken up into several smaller functions. It's about 150 lines with many conditionals, and thus it's difficult to understand all the things that comprise compiling match options. Perhaps it's an exercise for later, I just find long methods like this to be difficult to read and understand. 

I know it can be difficult too to create lots of small functions in XQuery. My experience has been that in the absence of classes (which are very effective at carrying out related bits of data with self-documenting field names), we can use maps and arbitrary XML fragments to carry around related bits of data. So a typical problem I've run into with trying to decompose a large XQuery function into smaller functions is I've got a lot of state represented by variables, and I may have to pass in several arguments to a function in order for it to do what it needs to do. So I do find it more difficult to create lots of small functions with self-documenting names that do one thing and one thing only - at least it's more difficult in XQuery than in Java. Still, I think at some point, we need more of this in the mastering code to help more developers be able to read and understand it, and thus be able to maintain and enhance it.